### ISSUE TITLE
Confusing diagnostics for lifetime on associated type

### ISSUE THREAD

**Issue by panstromek** [rust-lang/rust#76451]
When you declare this:
```rust
trait HasAssoc {
    type Assoc;
    fn tap(some: Self::Assoc) -> Self::Assoc;
}
impl HasAssoc for i32 {
    type Assoc = &'a i32;

    fn tap(some: Self::Assoc) -> Self::Assoc {
        some
    }
}
```
you get this error:
```
error[E0261]: use of undeclared lifetime name `'a`
  --> src\main.rs:12:19
   |
12 |     type Assoc = &'a i32;
   |                   ^^ undeclared lifetime
   |
help: consider introducing lifetime `'a` here
   |
11 | impl<'a> HasAssoc for i32 {

```
but if you follow the suggestion and place the lifetime parameter on the impl block, you get this:

```rust
error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
  --> src\main.rs:11:6
   |
11 | impl<'a> HasAssoc for i32 {
   |      ^^ unconstrained lifetime parameter
```

Until today, I thought this requires GATs and I wanted to suggest to put this information in the error message (which some other errors already tells you). But after playing around, I am actually not sure if it requires GATs or if it's even possible - either way, I don't know how to fix the error, so I think the diagnostics could be improved 😉

**Comment by panstromek** [comment]
Now..let's try if I can do this..
@rustbot modify labels : +A-diagnostics

**Comment by estebank** [comment]
[Not exactly pretty, ergonomic or stable](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8561d92cd591c4fab46a1a18db989ae7).

**Comment by panstromek** [comment]
Yes, I figured out both eventually, but both are not what I want 😄 It's been some time ago, but I recall that i needed to implement a trait from `std` that was not parametrized by a lifetime at all, so I was looking for a solution thad didn't require changing the trait definition but it would still allow having a reference as associated type.

**Comment by estebank** [comment]
That's a good point. In some suggestions we do check whether the involved trait comes from the current or an external crate and provide different suggestions for them. If in the current crate we can suggest changing the trait definition.

What did you end up using? Knowing that might help direct the general suggestion towards that.

**Comment by panstromek** [comment]
I think I did something different in the end. If I recall correctly, I wanted to return  an error from `FromStr` which would include the original `&str` (something like `MyParseError::InvalidString(&str)`). So I just changed it to `String`, which was better for providing full error message in my case, but boxing is probably not really a good suggestion for this problem in general.
